src/arch/riscv/isa.cc

```diff
@@ -985,6 +1018,250 @@ ISA::resetThread()
     Reset().invoke(tc);
 }
 
+bool
+ISA::isV() const
+{
+    // Virtualized when V-bit is 1
+    return readMiscRegNoEffect(MISCREG_VIRT) == 1;
+}
+
+void
+ISA::swapToVirtCSR(
+    uint64_t& csr, RegIndex& midx, std::string& csrName)
+{
+    switch (csr) {
+        case CSR_SSTATUS:  csr = CSR_VSSTATUS;  break;
+        case CSR_STVEC:    csr = CSR_VSTVEC;    break;
+        case CSR_SSCRATCH: csr = CSR_VSSCRATCH; break;
+        case CSR_SEPC:     csr = CSR_VSEPC;     break;
+        case CSR_SCAUSE:   csr = CSR_VSCAUSE;   break;
+        case CSR_STVAL:    csr = CSR_VSTVAL;    break;
+        case CSR_SATP:     csr = CSR_VSATP;     break;
+        case CSR_SIP:      csr = CSR_VSIP;      break;
+        case CSR_SIE:      csr = CSR_VSIE;      break;
+        default:           return;              break;
+    }
+
+    // Lookup new midx and csrName
+    auto csr_data_it = CSRData.find(csr);
+    if (csr_data_it == CSRData.end()) {
+        panic("Bad remapping of virtualized CSR");
+    }
+    midx = csr_data_it->second.physIndex;
+    csrName = csr_data_it->second.name;
+}
+
+Fault
+ISA::tvmChecks(uint64_t csr, PrivilegeMode pm, ExtMachInst machInst)
+{
+    if (csr == CSR_SATP) {
+        STATUS status = readMiscReg(MISCREG_STATUS);
+        if (pm != PRV_M && status.tvm == 1) {
+            return std::make_shared<IllegalInstFault>(
+                    "SATP access with TVM enabled\n",
+                    machInst);
+        }
+    }
+    else if (csr == CSR_VSATP) {
+        if (isV()) {
+        HSTATUS hstatus = readMiscReg(MISCREG_HSTATUS);
+        if (hstatus.vtvm == 1)
+            return std::make_shared<VirtualInstFault>(
+                "VSATP access with hstatus.vtvm enabled",
+                machInst);
+        }
+    }
+    else if (csr == CSR_HGATP) {
+        STATUS status = readMiscReg(MISCREG_STATUS);
+        if (pm != PRV_M && status.tvm == 1) {
+            return std::make_shared<IllegalInstFault>(
+                    "HGATP access with TVM enabled\n",
+                    machInst);
+        }
+    }
+
+
+    return NoFault;
+}
+
+RegVal
+ISA::backdoorReadCSRAllBits(uint64_t csr) {
+
+    auto csr_it = CSRData.find(csr);
+
+    // panic if the method was used with bad csr idx
+    panic_if(csr_it == CSRData.end(),
+        "Illegal CSR passed to backdoorReadCSRAllBits");
+
+    auto midx = csr_it->second.physIndex;
+
+    RegVal readval = readMiscReg(midx);
+
+    // Special handling for FCSR
+    if (csr == CSR_FCSR) {
+        readval = (readMiscReg(MISCREG_FFLAGS) |
+            (readMiscReg(MISCREG_FRM) << FRM_OFFSET));
+    }
+
+    return readval;
+}
+
+RegVal
+ISA::readCSR(uint64_t csr)
+{
+    auto csr_it = CSRData.find(csr);
```

Comment:
`getCSRDataMap()`? So that we can get customized CSR without panic

Ref: https://gem5-review.googlesource.com/c/public/gem5/+/62891